# Software Requirements Specification
Revision: v1.0.0

### Table Of Contents
1. Introduction
   1. Purpose
   2. Conventions
   3. Intended Audience
   4. Overall Description
2. System Features
3. Nonfunctional Requirements
4. Appendix

## Introduction
### Purpose
This project intends to provide capability for interpretation of genotype calls from the Mouse 
Universal Genotyping array platforms by building a web application that utilizes a programming 
interface to be provided by the client.

### Conventions
- WebApp - An application that runs through a web browser
- Module - An application component that can be used within other applications
- FR - Functional Requirement
- NFR - Nonfunctional Requirement

### Intended Audience
This document is intended for all clients, users and developers on the HaploQA project. 

### Overall Description
> HaploQA is a software platform that provides interpretation of genotype calls from the Mouse 
> Universal Genotyping array platforms (including Mega-/Giga- MUGA) that are available from Neogen 
> Genomics (Morgan et al., 2015). The software allows for direct upload of sample and genotype data,
> sample annotation, and organization by users. The software interprets genotype calls and uses 
> established Hidden Markov Models that were generated by Dr. Gary Churchillâ€™s laboratory to 
> generate haplotypes with negative log likelihood scoring. There are additional options for 
> interfacing with the data through a web-based graphics, links out to widely used genome browsers, 
> data/image/report downloads, and options for making the data public. If you want to read more 
> about MUGA platforms the following is a great resource:
>> Morgan AP et al.
>> The Mouse Universal Genotyping Array: From Substrains to Subspecies.
>> G3 (Bethesda). 2015 Dec 18;6(2):263-79. doi: 10.1534/g3.115.022087.
>> PMID: 26684931; PMCID: PMC4751547

### Product Description
We have been asked to design a web based application that utilizes a modular visualization design, 
to build these modular visualizations by migrating them from legacy HaploQA, and to start migrating 
the legacy frontend application pages.

### Product features
#### Web Application
The web application allows users to browse samples, view their metadata and visualizations, and to 
create an account and log in.

#### Visualization Modules
The visualizations provide an interactive way to view various data present about the sample on the
sample detail page. Modularizing these visualizations will allow for their reuse in additional
scientific applications.

### Client and Stakeholders

#### Clients
Dave Walton - The Jackson Laboratory<br>
Senior Principal Scientific Software Engineer<br>
Email: dave.walton@jax.org

Anna Lamoureux - The Jackson Laboratory<br>
Scientific Software Engineer<br>
Email: dave.walton@jax.org

#### Stakeholders
Laura Reinholdt - The Jackson Laboratory<br>
Associate Professor<br>
[Web Bio](https://www.jax.org/research-and-faculty/faculty/laura-reinholdt)

### System users
HaploQA is used by a wide variety of scientific and research focused users. It is assumed that the
users of HaploQA have a foundational technical understanding of the concepts presented.

### Operating environment
All contributions to HaploQA from this team are to be focused on the web application environment. 
The primary web browsers that are to be supported are to be in line with those supported by the
Angular web framework. These are, in order:
- Chrome: latest
- Firefox: latest and extended support release (ESR)
- Safari: 2 most recent major versions
- Edge: 2 most recent major versions

See the [Angular Documentation](https://angular.io/guide/browser-support#browser-support).

### Constraints
- The application needs to be developed in the Angular application framework
- The application needs to utilize a modular framework for visualizations
- The application needs to adhere to an API contract provided by the client

### User documentation requirements
- Documentation needs to be provided for both future software developers
  - This should include both code comments and descriptions in a relevant 
  README.md document.
- Documentation should be provided for end-users when necessary
  - Application end-users are assumed to understand the science, which does not
  need to be documented.

### Assumptions and dependencies
- The new programming interface to be used by this web application will be 
created by the clint.
- The system end users will have sufficient scientific understanding of the application 
functionality

## System Features 
The web application must provide the user with the following features:
- Sample List and Search
  - FR-1.0: Provide a list of all samples that can be viewed
  - FR-1.1: Provide a way to search by text for samples
  - FR-1.2: Provide a way to list search results
  - FR-1.3: Provide a way to list samples by genetic strain

- Sample Detail View
  - FR-2.0: Display sample metadata
    - FR-2.0.0: Page Title for Sample
    - FR-2.0.1: All additional metadata details provided by API contract
  - FR-2.1: Links to external browser resources
    - FR-2.1.0: Ensemble Browser
    - FR-2.1.1: UCSC Browser
  - FR-2.2: Provides visualizations
    - Visualizations provided through visualization module discussed below

- Accounts
  - FR-3.0: Ability to create a new account
  - FR-3.1: Ability to log in
  - FR-3.2: Allow accounts to provide information on affiliation with institutions

#### Visualization Modules
The visualization modules must provide the user (both developers and end-users) with the following
features:
- Visualizations
  - FR-4.0: Genome Karyotype Plot
  - FR-4.1: Genome Interval Plot

- Modularization method
  - FR-5.0: Can be installed into an Angular application
  - FR-5.1: Provides a way for the application to add its own data
  - FR-5.2: Provide module usage documentation

### Description and priority
#### High Priority
- All FR-5.X and one of FR-4.X
- FR-1.0, FR-1.1 and FR-1.2
- FR-2.2

#### Moderate Priority
- FR-1.3
- FR-2.0

#### Low Priority
- FR-2.1
- All of FR-3.X

## Nonfunctional Requirements
### Testing and Coverage
- NFR-1.0: Critical components should have automated testing written
- NFR-1.1: Automated tests should be run frequently to assure stability

### Modularization
- NFR-2.0: Documentation on module usage should be comprehensive and understandable
- NFR-2.1: It should be straightforward to use the module in an application

## Appendix
### Diagram One: 
    - TBD
### Diagram Two:
    - TBD
